global_match/hovering.py

```diff
@@ -0,0 +1,229 @@
+        G,
+        pos,
+        nodelist=rest,
+        node_color=colour2,
+        node_size=600,
+        node_shape=matplotlib.markers.MarkerStyle(
+            marker='o',
+            fillstyle='bottom'))
+    ''' 
+    Networkx by default draws straight arcs and places edge labels on the middle of those arcs.
+    However, we draw curved arcs but edge labels still remain at their default position (midpoint of NodeA and NodeB) {inside the cycle}
+    Thus we need to offset this by supplying new positions. To maintain consistency across all scales of X and Y axis, 
+    and positions of nodes we take the offset as 0.3 times difference between x-coordinates of the two nodes between which 
+    the edge is drawn. Different offsets are required for top edge and bottom edge of two cycles. For three cycles, the default 
+    placement causes no issue.
+     '''
+    pos_higher, pos_lower = {}, {}
+    # calculating offset
+    if len(top_edges) != 0:
+        a, b = top_edges[0]
+        y_off = 0.3 * abs(pos[a][0] - pos[b][0])
+    else:
+        y_off = 20
+
+    for k, v in pos.items():
+        pos_higher[k] = (v[0], v[1] + y_off)
+    for k, v in pos.items():
+        pos_lower[k] = (v[0], v[1] - y_off)
+    '''
+    w_top : edge weights of top edges of two cycles
+    w_bottom : edge weights of bottom edges of two cycles
+    w_rest : edge weights of remaining edges which can be placed in their default location
+    '''
+    w_top = {
+        e: str(
+            weight[e]) for e in weight if (
+            e in top_edges and e in G.edges())}
+    w_bottom = {
+        e: str(
+            weight[e]) for e in weight if (
+            e in bottom_edges and e in G.edges())}
+    w_rest = {
+        e: str(
+            weight[e]) for e in weight if (
+            e in G.edges() and e not in top_edges and e not in bottom_edges)}
+    ### Drawing edge labels
+    nx.draw_networkx_edges(
+        G,
+        pos,
+        edgelist=G.edges(),
+        connectionstyle='arc3,rad=0.2',
+        arrowsize=20)
+    nx.draw_networkx_edge_labels(
+        G,
+        pos_higher,
+        edge_labels=w_top,
+        label_pos=0.5,
+        verticalalignment='top')
+    nx.draw_networkx_edge_labels(
+        G,
+        pos_lower,
+        edge_labels=w_bottom,
+        label_pos=0.5,
+        verticalalignment='bottom')
+    nx.draw_networkx_edge_labels(G, pos, edge_labels=w_rest, label_pos=0.5)
+
+    # =================== HOVERING =========================
+    ### setting annotation style
+    annot = ax.annotate(
+        "", xy=(
+            0, 0), xytext=(
+            20, 20), textcoords="offset points", bbox=dict(
+                boxstyle="round", fc="w"), arrowprops=dict(
+                    arrowstyle="->"))
+    annot.set_visible(False)
+    idx_to_node_dict = {}
+    for idx, node in enumerate(G.nodes):
+        idx_to_node_dict[idx] = node
+        
+    ### setting annotation text
+    def update_annot(ind, nodelist):
```

Comment:
can you add some documentation using the python docstring? It is quite difficult to understand what the variable and function do.

Reply:
Sure I'll do that